components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/AbstractHTTPFunction.java

```diff
@@ -73,65 +82,242 @@ public AbstractHTTPFunction() {
+                    LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+                }
+                LOG.error("Received unknown response from external API call. Status code: " +
+                        responseCode + ". Url: " + endpointURL);
+                outcome = Constants.OUTCOME_FAIL;
+                return Pair.of(true, Pair.of(outcome, null)); // Server error, retry if attempts left
+            }
+        } catch (Exception e) {
+            // Log the error based on its type
+            if (e instanceof IllegalArgumentException) {
+                if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                    DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                            DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                            Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                    diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.API, endpointURL)
+                            .resultMessage("Invalid Url for external API call.")
+                            .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                            .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                    LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+                }
+                outcome = Constants.OUTCOME_FAIL;
+                LOG.error("Invalid Url: " + endpointURL, e);
+            } else if (e instanceof ConnectTimeoutException || e instanceof SocketTimeoutException) {
+                if (LoggerUtils.isDiagnosticLogsEnabled()) {
+                    DiagnosticLog.DiagnosticLogBuilder diagnosticLogBuilder = new
+                            DiagnosticLog.DiagnosticLogBuilder(Constants.LogConstants.ADAPTIVE_AUTH_SERVICE,
+                            Constants.LogConstants.ActionIDs.RECEIVE_API_RESPONSE);
+                    diagnosticLogBuilder.inputParam(Constants.LogConstants.InputKeys.API, endpointURL)
+                            .resultMessage("Request for the external API timed out.")
+                            .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                            .resultStatus(DiagnosticLog.ResultStatus.FAILED);
+                    LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
+                }
+                isRetry = true; // Timeout, retry if attempts left
                 outcome = Constants.OUTCOME_TIMEOUT;
-            } catch (SocketTimeoutException e) {
-                LOG.error("Error while waiting for data from " + endpointURL, e);
+                LOG.error("Error while waiting to connect to " + endpointURL, e);
+            } else if (e instanceof IOException) {
+                isRetry = true; // Timeout, retry if attempts left
                 outcome = Constants.OUTCOME_TIMEOUT;
-            } catch (IOException e) {
                 LOG.error("Error while calling endpoint. ", e);
```

Comment:
I don't think this should be marked as a time out.
This error happens due to a failure in the system when acting as a client to invoke the respective API. Thus, this should not be considered as a time out scenario and retrying the request is not needed IMO

Reply:
Resolved with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/01b5e2f5f849d7a4e89d792e28845597544fca39